litex/soc/interconnect/csr.py

```diff
@@ -122,6 +122,66 @@ def do_finalize(self, busword):
         raise NotImplementedError
 
 
+class CSRField(Signal):
+    def __init__(self, name, size=1, offset=None, reset=0, description=None, pulse=False, access=None, values=None):
+        assert name == name.lower()
+        assert access in [None, "write-only", "read-only", "read-write"]
+        self.name        = name
+        self.size        = size
+        self.offset      = offset
+        self.reset_value = reset
+        self.description = description
+        self.access      = access
+        self.pulse       = pulse
+        self.values      = values
+        Signal.__init__(self, size, name=name, reset=reset)
+
+
+class CSRFieldCompound:
+    def __init__(self, fields, access):
+        self.check_names(fields)
+        self.check_ordering_overlap(fields)
+        self.fields = fields
+        for field in fields:
+            if field.access is None:
+                field.access = access
+            elif access == "read-only":
+                assert field.access == "read-only"
+            elif access == "read-write":
+                assert field.access in ["read-write", "write-only"]
+            setattr(self, field.name, field)
+
+    @staticmethod
+    def check_names(fields):
+        names = []
+        for field in fields:
+            if field.name in names:
```

Comment:
Should we make sure names are lowercase? It is possible to make `CSRField("name")` and `CSRField("Name")`, which won't violate a check here but will generate a CPP error in the resulting `csr.h` file.

Reply:
The check is done directly in the field: https://github.com/enjoy-digital/litex/pull/257/commits/8c080e5fb6cf52342e4e4b5963d9d85e2e90a57a#diff-ff1bf16eefbf92d8e4569d807b041554R127 but indeed, we it would be more consistent to move it to check_names. We are converting the name to uppercase when generating the `csr.h`, so in fact i'm not sure this is really useful. I'll check that.

Reply:
Ah, that's what `lower()` does.  That works, then!